src/libraries/System.Private.CoreLib/src/System/Int16.cs

```diff
@@ -125,102 +126,44 @@ bool IUtf8SpanFormattable.TryFormat(Span<byte> utf8Destination, out int bytesWri
             return Number.TryFormatInt32(m_value, 0x0000FFFF, format, provider, utf8Destination, out bytesWritten);
         }
 
-        public static short Parse(string s)
-        {
-            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
-            return Parse((ReadOnlySpan<char>)s, NumberStyles.Integer, NumberFormatInfo.CurrentInfo);
-        }
+        public static short Parse(string s) => Parse(s, NumberStyles.Integer, provider: null);
 
-        public static short Parse(string s, NumberStyles style)
-        {
-            NumberFormatInfo.ValidateParseStyleInteger(style);
-            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
-            return Parse((ReadOnlySpan<char>)s, style, NumberFormatInfo.CurrentInfo);
-        }
+        public static short Parse(string s, NumberStyles style) => Parse(s, style, provider: null);
 
-        public static short Parse(string s, IFormatProvider? provider)
-        {
-            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
-            return Parse((ReadOnlySpan<char>)s, NumberStyles.Integer, NumberFormatInfo.GetInstance(provider));
-        }
+        public static short Parse(string s, IFormatProvider? provider) => Parse(s, NumberStyles.Integer, provider);
 
         public static short Parse(string s, NumberStyles style, IFormatProvider? provider)
         {
-            NumberFormatInfo.ValidateParseStyleInteger(style);
-            if (s == null) ThrowHelper.ThrowArgumentNullException(ExceptionArgument.s);
-            return Parse((ReadOnlySpan<char>)s, style, NumberFormatInfo.GetInstance(provider));
+            ArgumentNullException.ThrowIfNull(s);
```

Comment:
In most of corelib on hot paths and where we have a ThrowHelper that works, we use it, since the code ends up being a bit tighter.  Why replace it in these files?

Reply:
Removes code duplication and ensures the exceptions thrown are generally consistent with what most users will see.

The JIT does the "right stuff" for the throw helpers, so it seems wasteful to not use them.

Reply:
> Removes code duplication

I'm not sure what you mean. Do you just mean shorter? 

> and ensures the exceptions thrown are generally consistent with what most users will see

The exception thrown is the same type with the same argument. Are you referring to something else?

> The JIT does the "right stuff" for the throw helpers, so it seems wasteful to not use them.

The asm is still larger with ThrowIfNull.
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGBLAHYYGAWQAUxcgAYGuAJRcenaj1UMAglADmAV3wwhAOR0AbEwFEEYGAAcMfCAIB0AFQAWUCAHcAkgDNjMzF5BgBuRTUmAHZZJwAZAy0MN3CVHgBfCIYs+gZBYRFSCWlZOWy07mVI7j4/BmC83AYBCGEBUxMy908vAAkYExsYKFcPb00tQJMxbu9+weGnCdwnENTq6NiEgSSUiPTyzOoaZlm+gaGoGiqeXIM9DW0m9lkj1RzGSSQmFAYziamYmWDGwZQAvAA+LKqZI9ZowLyPXT6IwdSzWOwOARiADiMAwhmw+jEoLKqSyNigfAAbtgMDAmORvpIZHiCUSYECniDwRCQa4IABlDBUnZiOSpA5AA=) 
And while we generally care more about that with generics and these aren't, these callers are also going to be inlined into lots of places. 

Unless we're planning to rip them all out, or even just all of them that use "s", I'm not sure why we're doing so here.

Reply:
For me, this really boils down to we should be using the APIs we ship for the broader public. If these throw helpers weren't actually sufficient for use in our hot paths, then we likely should have an explicit bug tracking whatever underlying issue exists and working towards fixing that.

I don't see such issue, however. The hot path logic is exactly the same, the throwing path is slightly larger because the logic to load the string literal exists before the `Throw()` call rather than in the `Throw()` call. The throwing path is "cold" and so it doesn't actually execute and this slightly larger assembly won't be impactful.

For IL, the method bodies differ in that `ExceptionArguments.s` takes 1-more IL instruction, but 1-less byte IL. Neither of these points will substantially impact what the JIT can or has to do.

Planned improvements to the handling of cold blocks in the JIT will help here in the future as well, for both scenarios.

Reply:
These APIs are used in orders of magnitude more places than others, such that even a small amount of size impact can matter.

If there's truly zero downside here to such a change, for coreclr and mono, then we should instead be systematically replacing all such ThrowArgumentNull usage and deleting the supporting helpers. If there isn't, then we shouldn't be one-off removing them here.

Reply:
> These APIs are used in orders of magnitude more places than others, such that even a small amount of size impact can matter.

Yes, IL size can matter. The 1-byte additional IL won't matter nor will the 1 instruction less.

> If there's truly zero downside here to such a change, for coreclr and mono, then we should instead be systematically replacing all such ThrowArgumentNull usage and deleting the supporting helpers. If there isn't, then we shouldn't be one-off removing them here.

I agree, and we should be logging bugs for anything that would block its usage, if anything does come up. I'm not currently aware of anything that would cause it to be blocked and don't see any hot path diffs in an SPMI comparison either.

Reply:
> Yes, IL size can matter.

And asm size.

> we should be logging bugs for anything that would block its usage

Let me try again. I don't think we should be removing these one-off. If there's no benefit to them, please submit a separate pr cleaning them up everywhere. If there's still benefit, they shouldn't be removed here. Either way, they shouldn't be removed here.

Reply:
> And asm size.

Asm size does not impact the JIT's ability to optimize or anything like that, it only impacts the resulting size of the generated library/exe in the case of crossgen. Larger assembly does not mean slower code, it can often be quite the opposite and much of the compilers job is avoiding slow "macro" instructions in favor of alternative "micro-op" sequences instead. In this case, the additional assembly only exists on the slow path and is overall less total assembly to execute than we had before.

> Let me try again. I don't think we should be removing these one-off. If there's no benefit to them, please submit a separate pr cleaning them up everywhere. If there's still benefit, they shouldn't be removed here. Either way, they shouldn't be removed here.

This code was being generally cleaned up and simplified to avoid duplication and improve maintainability. Fixing up the argument validation to use the throw helpers fits in perfectly with those types of changes. I don't see the reason to churn CI again to pull out a perfectly valid change where perf numbers have already shown it doesn't have a negative impact on RyuJIT and that the general change is likewise either an improvement or within the acceptable margin for regression on Mono.

Reply:
If I understand this debate correctly:
- We're not yet confident that the new public throw helpers should be used in the lowest layers and hottest paths
  - Using the public throw helpers keeps the code more terse and consistent throughout our repo
  - But they yield more IL, and for a method this low and commonly used, that can compound in measurable ways
- The spirit of this PR was the deduplicate code and modernize it a bit
  - Replacing the internal throw helper usage with the public throw helper usage seems apropos
  - But a concern was raised that making such changes opportunistically (here and in other PRs) could inadvertently hide aggregated size impacts of taking this approach (piece by piece)

I think we then have 2 different positions, both of which are valid:
1. Making the code more terse and consistent is a good goal and the changes were in the spirit of this PR, and the impact of this singular change is uncertain
2. Switching from our internal throw helpers to the public ones in the lowest layers and hottest paths will likely have an aggregate impact on assembly size reduction goals, and we should therefore test that approach before adopting it

Given that we do have a heavy investment into our total assembly sizes across several scenarios (with several team members currently devoted to that goal), and this path is certainly used in those scenarios, I too think we should revert this throw helper change (and the other in `Char.Parse`).

If we want to test the impact of switching from internal throw helpers to the public ones throughout the entire repo, we could file an issue and submit a PR that does exclusively that. We could then measure the impact of that aggregate direction through the perf lab where we could also see how the assembly size impacts add up.

Reply:
> But they yield more IL, and for a method this low and commonly used, that can compound in measurable ways

More IL is a bit subjective. It is 1 IL instruction less but 1 byte more in terms of total IL size.

> But a concern was raised that making such changes opportunistically (here and in other PRs) could inadvertently hide aggregated size impacts of taking this approach (piece by piece)

We are already using the `ArgumentNullException.ThrowIfNull` helper in over 8000 places throughout the BCL. We only didn't take it in a few selective spots, such as these "very hot methods".

The vast majority of the codebase was updated to use it back in April of last year (https://github.com/dotnet/runtime/pull/68178) and we saw no evidence of any negative perf impact in our perf triage since (we actually did see improvements in methods that switched to using any helper where they previously threw directly). Nor did we see any real increase in crossgen size from such changes:
![image](https://user-images.githubusercontent.com/10487869/232331771-c2218511-1803-48e2-914c-322ddb9a01e1.png)

Reply:
> The vast majority of the codebase was updated to use it back in April of last year

That change didn't touch use of ThrowHelper.  It changed use of `!!`, and the change that switched to use `!!` also didn't touch direct use of ThrowHelper.  Those changes were all about code that was manually instantiating/throwing exceptions.

> This PR provides perf numbers showing that there is not a negative perf impact

It provides throughput numbers.  I've primarily commented on asm size.  We do have efforts that care about asm size, e.g. native AOT.

My request has been simple. If ThrowHelper.ThrowArgumentNullException(ExceptionArgument) is antiquated, then let's remove them all in a single PR rather than doing it 10 here and 10 there (it's used ~500 times in corelib), while also removing the ThrowArgumentNullException helper itself and all the enum members used just for it.

Reply:
> If ThrowHelper.ThrowArgumentNullException(ExceptionArgument) is antiquated, then let's remove them all in a single PR rather than doing it 10 here and 10 there (it's used ~500 times in corelib), while also removing the ThrowArgumentNullException helper itself and all the enum members used just for it.

That's what I'd like to see as well. Making a deliberate and sweeping adoption of the new helpers while under specific measurement seems prudent to me. We can then measure the total effects from each angle.

I appreciate where you're coming from @tannergooding, and I see how this change seems appropriate while in the code here already. I think we just haven't written down the DOs and the DON'Ts of when/where to use these helpers yet.

Reply:
> My request has been simple. If ThrowHelper.ThrowArgumentNullException(ExceptionArgument) is antiquated, then let's remove them all in a single PR rather than doing it 10 here and 10 there (it's used ~500 times in corelib), while also removing the ThrowArgumentNullException helper itself and all the enum members used just for it.

And I'm more than happy to put up a PR doing that.

I don't see why that requires this PR to undo the change and rerun CI to happen. To me, that seems like needless additional churn particularly when there are benefits to merging it as is before the a broader PR goes up. For example, it means we are able to get proper numbers in the perf lab against some of the most primitive functions, which could in turn influence whether or not anything unexpected may crop up from such a broader change.

Reply:
I'd rather test the effects in aggregate and in isolation so that we can see how it accumulates. With all of the changes in this PR, the effects would get dilluted here and not give a strong enough signal to base our data on. The CI churn is worthwhile to me for us to have a clear signal whether or not we should remove all of the internal throw helpers and use the public ones, or if there are proven cases where it makes sense to keep using the internal ones.